From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bruno de Lacheisserie <bruno.de.lacheisserie@gmail.com>
Date: Fri, 29 Mar 2024 01:21:42 +0100
Subject: [PATCH] get_cert_by_subject : backward compatibility

Fallback to MD5 name hash if no certificate found using SHA1 hash

fixes #13565 #15154

Upstream: https://github.com/openssl/openssl/pull/24002

This squashes the six commits of that pull request and rebases them onto
OpenSSL 3.5.9, whose by_dir.c was reformatted with clang-format and now
skips broken symlinks and certificates that fail to load. The lookup loop
from 3.5.9 is kept as is and wrapped in the loop over both name hashes.
---
 crypto/x509/by_dir.c                  | 295 ++++++++++++++------------
 test/recipes/60-test_x509_store_old.t |  53 +++++
 2 files changed, 208 insertions(+), 140 deletions(-)
 create mode 100644 test/recipes/60-test_x509_store_old.t

diff --git a/crypto/x509/by_dir.c b/crypto/x509/by_dir.c
index df92e1d..e1d1b25 100644
--- a/crypto/x509/by_dir.c
+++ b/crypto/x509/by_dir.c
@@ -229,7 +229,9 @@ static int get_cert_by_subject_ex(X509_LOOKUP *xl, X509_LOOKUP_TYPE type,
         X509_CRL crl;
     } data;
     int res, ok = 0;
-    int i, j, k;
+    int i, j, k, l;
+    unsigned long hash[2];
+    int valid_hash[2]; /* '0' for invalid value in `hash` array */
     unsigned long h;
     BUF_MEM *b = NULL;
     X509_OBJECT stmp, *tmp;
@@ -257,175 +259,188 @@ static int get_cert_by_subject_ex(X509_LOOKUP *xl, X509_LOOKUP_TYPE type,
     }
 
     ctx = (BY_DIR *)xl->method_data;
-    h = X509_NAME_hash_ex(name, libctx, propq, &i);
-    if (i == 0)
-        goto finish;
-    for (i = 0; i < sk_BY_DIR_ENTRY_num(ctx->dirs); i++) {
-        BY_DIR_ENTRY *ent;
-        int idx;
-        BY_DIR_HASH htmp, *hent;
-
-        ent = sk_BY_DIR_ENTRY_value(ctx->dirs, i);
-        j = strlen(ent->dir) + 1 + 8 + 6 + 1 + 1;
-        if (!BUF_MEM_grow(b, j)) {
-            ERR_raise(ERR_LIB_X509, ERR_R_BUF_LIB);
-            goto finish;
-        }
-        if (type == X509_LU_CRL && ent->hashes) {
-            htmp.hash = h;
-            if (!CRYPTO_THREAD_read_lock(ctx->lock))
+
+    hash[0] = X509_NAME_hash_ex(name, libctx, propq, &l); /* SHA1 name hash */
+    valid_hash[0] = (l != 0);
+#ifndef OPENSSL_NO_MD5
+    hash[1] = X509_NAME_hash_old(name); /* MD5 name hash */
+    valid_hash[1] = 1;
+#else
+    valid_hash[1] = 0;
+#endif
+
+    for (l = 0; l < 2; ++l) {
+        h = hash[l];
+        if (valid_hash[l] == 0)
+            continue; /* skip invalid hash */
+
+        for (i = 0; i < sk_BY_DIR_ENTRY_num(ctx->dirs); i++) {
+            BY_DIR_ENTRY *ent;
+            int idx;
+            BY_DIR_HASH htmp, *hent;
+
+            ent = sk_BY_DIR_ENTRY_value(ctx->dirs, i);
+            j = strlen(ent->dir) + 1 + 8 + 6 + 1 + 1;
+            if (!BUF_MEM_grow(b, j)) {
+                ERR_raise(ERR_LIB_X509, ERR_R_BUF_LIB);
                 goto finish;
-            idx = sk_BY_DIR_HASH_find(ent->hashes, &htmp);
-            if (idx >= 0) {
-                hent = sk_BY_DIR_HASH_value(ent->hashes, idx);
-                k = hent->suffix;
+            }
+            if (type == X509_LU_CRL && ent->hashes) {
+                htmp.hash = h;
+                if (!CRYPTO_THREAD_read_lock(ctx->lock))
+                    goto finish;
+                idx = sk_BY_DIR_HASH_find(ent->hashes, &htmp);
+                if (idx >= 0) {
+                    hent = sk_BY_DIR_HASH_value(ent->hashes, idx);
+                    k = hent->suffix;
+                } else {
+                    hent = NULL;
+                    k = 0;
+                }
+                CRYPTO_THREAD_unlock(ctx->lock);
             } else {
-                hent = NULL;
                 k = 0;
+                hent = NULL;
             }
-            CRYPTO_THREAD_unlock(ctx->lock);
-        } else {
-            k = 0;
-            hent = NULL;
-        }
-        for (;;) {
-            char c = '/';
+            for (;;) {
+                char c = '/';
 
 #ifdef OPENSSL_SYS_VMS
-            c = ent->dir[strlen(ent->dir) - 1];
-            if (c != ':' && c != '>' && c != ']') {
-                /*
-                 * If no separator is present, we assume the directory
-                 * specifier is a logical name, and add a colon.  We really
-                 * should use better VMS routines for merging things like
-                 * this, but this will do for now... -- Richard Levitte
-                 */
-                c = ':';
-            } else {
-                c = '\0';
-            }
+                c = ent->dir[strlen(ent->dir) - 1];
+                if (c != ':' && c != '>' && c != ']') {
+                    /*
+                     * If no separator is present, we assume the directory
+                     * specifier is a logical name, and add a colon.  We really
+                     * should use better VMS routines for merging things like
+                     * this, but this will do for now... -- Richard Levitte
+                     */
+                    c = ':';
+                } else {
+                    c = '\0';
+                }
 
-            if (c == '\0') {
-                /*
-                 * This is special.  When c == '\0', no directory separator
-                 * should be added.
-                 */
-                BIO_snprintf(b->data, b->max,
-                    "%s%08lx.%s%d", ent->dir, h, postfix, k);
-            } else
+                if (c == '\0') {
+                    /*
+                     * This is special.  When c == '\0', no directory separator
+                     * should be added.
+                     */
+                    BIO_snprintf(b->data, b->max,
+                        "%s%08lx.%s%d", ent->dir, h, postfix, k);
+                } else
 #endif
-            {
-                BIO_snprintf(b->data, b->max,
-                    "%s%c%08lx.%s%d", ent->dir, c, h, postfix, k);
-            }
+                {
+                    BIO_snprintf(b->data, b->max,
+                        "%s%c%08lx.%s%d", ent->dir, c, h, postfix, k);
+                }
 #ifndef OPENSSL_NO_POSIX_IO
 #ifdef _WIN32
 #define lstat _stat
 #define stat _stat
 #endif
-            {
-                struct stat st;
-                if (lstat(b->data, &st) < 0)
-                    break; /* file does not exist, not even a symlink */
+                {
+                    struct stat st;
+                    if (lstat(b->data, &st) < 0)
+                        break; /* file does not exist, not even a symlink */
 #ifndef _WIN32
-                if (stat(b->data, &st) < 0) {
-                    k++;
-                    continue; /* symlink is broken: following it went wrong */
-                }
+                    if (stat(b->data, &st) < 0) {
+                        k++;
+                        continue; /* symlink is broken: following it went wrong */
+                    }
 #endif
-            }
+                }
 #endif
-            res = 0;
-            ERR_set_mark();
-            if (type == X509_LU_X509)
-                res = X509_load_cert_file_ex(xl, b->data, ent->dir_type, libctx, propq);
-            else if (type == X509_LU_CRL)
-                res = X509_load_crl_file(xl, b->data, ent->dir_type);
-            /* else case will caught higher up */
-            ERR_pop_to_mark();
-            /* unless OPENSSL_NO_POSIX_IO, gracefully skip found file if cert/CRL fails to load. */
+                res = 0;
+                ERR_set_mark();
+                if (type == X509_LU_X509)
+                    res = X509_load_cert_file_ex(xl, b->data, ent->dir_type, libctx, propq);
+                else if (type == X509_LU_CRL)
+                    res = X509_load_crl_file(xl, b->data, ent->dir_type);
+                /* else case will caught higher up */
+                ERR_pop_to_mark();
+                /* unless OPENSSL_NO_POSIX_IO, gracefully skip found file if cert/CRL fails to load. */
 #ifndef OPENSSL_NO_POSIX_IO
-            res = 1;
+                res = 1;
 #endif
-            if (res == 0)
-                break;
-            k++;
-        }
+                if (res == 0)
+                    break;
+                k++;
+            }
 
-        /*
-         * we have added it to the cache so now pull it out again
-         *
-         * Note: quadratic time find here since the objects won't generally be
-         *       sorted and sorting the would result in O(n^2 log n) complexity.
-         */
-        if (k > 0) {
-            if (!X509_STORE_lock(xl->store_ctx))
-                goto finish;
-            j = sk_X509_OBJECT_find(xl->store_ctx->objs, &stmp);
-            tmp = sk_X509_OBJECT_value(xl->store_ctx->objs, j);
-            X509_STORE_unlock(xl->store_ctx);
-        } else {
-            tmp = NULL;
-        }
-        /*
-         * If a CRL, update the last file suffix added for this.
-         * We don't need to add an entry if k is 0 as this is the initial value.
-         * This avoids the need for a write lock and sort operation in the
-         * simple case where no CRL is present for a hash.
-         */
-        if (type == X509_LU_CRL && k > 0) {
-            if (!CRYPTO_THREAD_write_lock(ctx->lock))
-                goto finish;
             /*
-             * Look for entry again in case another thread added an entry
-             * first.
+             * we have added it to the cache so now pull it out again
+             *
+             * Note: quadratic time find here since the objects won't generally be
+             *       sorted and sorting the would result in O(n^2 log n) complexity.
              */
-            if (hent == NULL) {
-                htmp.hash = h;
-                idx = sk_BY_DIR_HASH_find(ent->hashes, &htmp);
-                hent = sk_BY_DIR_HASH_value(ent->hashes, idx);
+            if (k > 0) {
+                if (!X509_STORE_lock(xl->store_ctx))
+                    goto finish;
+                j = sk_X509_OBJECT_find(xl->store_ctx->objs, &stmp);
+                tmp = sk_X509_OBJECT_value(xl->store_ctx->objs, j);
+                X509_STORE_unlock(xl->store_ctx);
+            } else {
+                tmp = NULL;
             }
-            if (hent == NULL) {
-                hent = OPENSSL_malloc(sizeof(*hent));
-                if (hent == NULL) {
-                    CRYPTO_THREAD_unlock(ctx->lock);
-                    ok = 0;
+            /*
+             * If a CRL, update the last file suffix added for this.
+             * We don't need to add an entry if k is 0 as this is the initial value.
+             * This avoids the need for a write lock and sort operation in the
+             * simple case where no CRL is present for a hash.
+             */
+            if (type == X509_LU_CRL && k > 0) {
+                if (!CRYPTO_THREAD_write_lock(ctx->lock))
                     goto finish;
+                /*
+                 * Look for entry again in case another thread added an entry
+                 * first.
+                 */
+                if (hent == NULL) {
+                    htmp.hash = h;
+                    idx = sk_BY_DIR_HASH_find(ent->hashes, &htmp);
+                    hent = sk_BY_DIR_HASH_value(ent->hashes, idx);
                 }
-                hent->hash = h;
-                hent->suffix = k;
-                if (!sk_BY_DIR_HASH_push(ent->hashes, hent)) {
-                    CRYPTO_THREAD_unlock(ctx->lock);
-                    OPENSSL_free(hent);
-                    ERR_raise(ERR_LIB_X509, ERR_R_CRYPTO_LIB);
-                    ok = 0;
-                    goto finish;
+                if (hent == NULL) {
+                    hent = OPENSSL_malloc(sizeof(*hent));
+                    if (hent == NULL) {
+                        CRYPTO_THREAD_unlock(ctx->lock);
+                        ok = 0;
+                        goto finish;
+                    }
+                    hent->hash = h;
+                    hent->suffix = k;
+                    if (!sk_BY_DIR_HASH_push(ent->hashes, hent)) {
+                        CRYPTO_THREAD_unlock(ctx->lock);
+                        OPENSSL_free(hent);
+                        ERR_raise(ERR_LIB_X509, ERR_R_CRYPTO_LIB);
+                        ok = 0;
+                        goto finish;
+                    }
+
+                    /*
+                     * Ensure stack is sorted so that subsequent sk_BY_DIR_HASH_find
+                     * will not mutate the stack and therefore require a write lock.
+                     */
+                    sk_BY_DIR_HASH_sort(ent->hashes);
+                } else if (hent->suffix < k) {
+                    hent->suffix = k;
                 }
 
-                /*
-                 * Ensure stack is sorted so that subsequent sk_BY_DIR_HASH_find
-                 * will not mutate the stack and therefore require a write lock.
-                 */
-                sk_BY_DIR_HASH_sort(ent->hashes);
-            } else if (hent->suffix < k) {
-                hent->suffix = k;
+                CRYPTO_THREAD_unlock(ctx->lock);
             }
 
-            CRYPTO_THREAD_unlock(ctx->lock);
-        }
-
-        if (tmp != NULL) {
-            ok = 1;
-            ret->type = tmp->type;
-            memcpy(&ret->data, &tmp->data, sizeof(ret->data));
+            if (tmp != NULL) {
+                ok = 1;
+                ret->type = tmp->type;
+                memcpy(&ret->data, &tmp->data, sizeof(ret->data));
 
-            /*
-             * Clear any errors that might have been raised processing empty
-             * or malformed files.
-             */
-            ERR_clear_error();
+                /*
+                 * Clear any errors that might have been raised processing empty
+                 * or malformed files.
+                 */
+                ERR_clear_error();
 
-            goto finish;
+                goto finish;
+            }
         }
     }
 finish:
diff --git a/test/recipes/60-test_x509_store_old.t b/test/recipes/60-test_x509_store_old.t
new file mode 100644
index 0000000..face5c7
--- /dev/null
+++ b/test/recipes/60-test_x509_store_old.t
@@ -0,0 +1,53 @@
+#! /usr/bin/env perl
+# Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+#
+# Licensed under the Apache License 2.0 (the "License").  You may not use
+# this file except in compliance with the License.  You can obtain a copy
+# in the file LICENSE in the source distribution or at
+# https://www.openssl.org/source/license.html
+
+
+use strict;
+use warnings;
+
+use File::Copy;
+use File::Spec::Functions qw/:DEFAULT canonpath/;
+use OpenSSL::Test qw/:DEFAULT srctop_file/;
+
+setup("test_x509_store_old");
+
+#If "openssl rehash -help" fails it's most likely because we're on a platform
+#that doesn't support the rehash command (e.g. Windows)
+plan skip_all => "test_rehash is not available on this platform"
+    unless run(app(["openssl", "rehash", "-help"]));
+
+# We use 'openssl verify' for these tests, as it contains everything
+# we need to conduct these tests.  The tests here are a subset of the
+# ones found in 25-test_verify.t
+
+sub verify {
+    my ($cert, $purpose, $trustedpath, $untrusted, @opts) = @_;
+    my @args = qw(openssl verify -auth_level 1 -purpose);
+    my @path = qw(test certs);
+    push(@args, "$purpose", @opts);
+    push(@args, "-CApath", $trustedpath);
+    for (@$untrusted) { push(@args, "-untrusted", srctop_file(@path, "$_.pem")) }
+    push(@args, srctop_file(@path, "$cert.pem"));
+    run(app([@args]));
+}
+
+plan tests => 3;
+
+indir "60-test_x509_store_old" => sub {
+    for (("root-cert")) {
+        copy(srctop_file("test", "certs", "$_.pem"), curdir());
+    }
+    ok(run(app([qw(openssl rehash -old), curdir()])), "Rehashing");
+
+    # Canonical success
+    ok(verify("ee-cert", "sslserver", curdir(), ["ca-cert"], "-show_chain"),
+       "verify ee-cert");
+
+    # Failure because root cert not present in CApath
+    ok(!verify("ca-root2", "any", curdir(), [], "-show_chain"));
+}, create => 1, cleanup => 1;
-- 
2.53.0

